dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/DefaultTrackerValidationService.java

```diff
@@ -28,53 +28,114 @@
+
+        if ( !VALIDATION_ORDER
+            .containsAll(
+                validationHooks.stream().map( TrackerValidationHook::getClass ).collect( Collectors.toList() ) ) )
+        {
+            String orderList = VALIDATION_ORDER.stream().map( Class::getName )
+                .collect( Collectors.joining( "," ) );
+
+            String internList = validationHooks.stream().map( i -> i.getClass().getName() )
+                .collect( Collectors.joining( "," ) );
+
+            throw new RuntimeException(
+                String.format(
+                    "ValidationConfig.class is missing a validation hook in the validation order list, " +
+                        "please add it to the list. Order list: %s, service list: %s",
+                    orderList, internList ) );
+        }
+
+        this.validationHooks.sort( Comparator.comparingInt( o -> VALIDATION_ORDER_MAP.get( o.getClass() ) ) );
     }
 
     @Override
     public TrackerValidationReport validate( TrackerBundle bundle )
     {
         TrackerValidationReport validationReport = new TrackerValidationReport();
 
-        if ( (bundle.getUser() == null || bundle.getUser().isSuper()) && ValidationMode.SKIP == bundle.getValidationMode() )
+        User user = bundle.getUser();
+
+        if ( (user == null || user.isSuper()) && ValidationMode.SKIP == bundle.getValidationMode() )
         {
-            log.warn( "Skipping validation for metadata import by user '" + bundle.getUsername() + "'. Not recommended." );
+            log.warn( "Skipping validation for metadata import by user '" +
+                bundle.getUsername() + "'. Not recommended." );
             return validationReport;
         }
 
-        for ( TrackerValidationHook hook : validationHooks )
-        {
-            validationReport.add( hook.validate( bundle ) );
+        TrackerImportValidationContext context = new TrackerImportValidationContext( bundle );
 
-            if ( !validationReport.isEmpty() && ValidationMode.FAIL_FAST == bundle.getValidationMode() )
+        try
+        {
+            for ( TrackerValidationHook hook : validationHooks )
             {
-                break;
+                if ( hook.isEnabled() )
+                {
+                    Timer hookTimer = Timer.startTimer();
+
+                    List<TrackerErrorReport> errors = hook.validate( context );
+                    if ( !errors.isEmpty() )
+                    {
+                        validationReport.add( errors );
+                    }
+
+                    validationReport.add( TrackerValidationHookTimerReport.builder()
+                        .name( hook.getClass().getName() )
+                        .totalTime( hookTimer.toString() ).build() );
+                }
             }
         }
+        catch ( ValidationFailFastException e )
+        {
+            validationReport.add( e.getErrors() );
+        }
+
+        if ( validationReport.getErrorReports().size() > 0 )
```

Comment:
You can remove this if.
Print report will do nothing if there are not error reports.
In general, if you need to do something with the elements on the list, you don't need to check the size